File: hw4/task1.py
import sys
from pyspark import SparkContext
import os
import time
from itertools import combinations
from pyspark.sql import SparkSession
from graphframes import GraphFrame


if __name__ == "__main__":

    #check the validation of the command
    if len(sys.argv) != 4:
        print("There is an error for the Usage: python3 task1.py <threshold> <input_filepath> <output_filepath>")
        sys.exit(1)

    sc = SparkContext(appName="Task1")

    thr = int(sys.argv[1])
    inputs = sys.argv[2]
    output = sys.argv[3]

    start_time = time.time()
    sc.setLogLevel("ERROR")
    csv_rdd = sc.textFile(inputs)
    header = csv_rdd.first()
    data_nh = (csv_rdd.filter(lambda x: x != header).map(lambda x: x.split(","))
               .map(lambda x: (x[0], {x[1]}))
               .reduceByKey(lambda a, b: a.union(b)))

    data_pre = data_nh.filter(lambda x: len(x[1]) >= thr).collectAsMap()

    uni_user = list(data_pre.keys())
    combuser = list(combinations(uni_user, 2))

    filter_p = sc.parallelize(combuser).filter(lambda users: len(data_pre[users[0]].intersection(data_pre[users[1]])) >= thr)
    edges_n = filter_p.flatMap(lambda users: [(users[1], users[0]), (users[0], users[1])])
    nodes = filter_p.flatMap(lambda users: [users[0], users[1]]).distinct().map(lambda user: (user,))

    spark = SparkSession.builder.appName("community").getOrCreate()
    vertices = spark.createDataFrame(nodes, ["id"])
    edges = spark.createDataFrame(edges_n, ["src", "dst"])
    g = GraphFrame(vertices, edges)
    result = g.labelPropagation(maxIter=5)
    final = result.rdd.map(lambda x: (x['label'], [x['id']]))
    final = final.reduceByKey(lambda a, b: a + b).map(lambda x: (sorted(x[1]), len(x[1]))).sortBy(lambda x: (x[1], x[0][0]))
    fin = final.map(lambda x: x[0])

    with open(output, 'w') as file:
        for f in fin.collect():
            com = ', '.join([f"'{u}'" for u in f])
            file.write(com + "\n")
    spark.stop()

    end_time = time.time()
    execution = end_time - start_time
    print(f"Duration: {execution}")
    sc.stop()

File: hw4/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.types.{StructType, StructField, StringType}
import org.apache.spark.graphx._
import org.apache.spark.sql.functions._
import org.graphframes.GraphFrame
import scala.io.Source
import java.io.{PrintWriter, File}

object task1 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("Task1")
    val sc = new SparkContext(conf)
    val spark = SparkSession.builder
      .appName("community")
      .getOrCreate()
    sc.setLogLevel("ERROR")
    spark.sparkContext.setLogLevel("ERROR")

    import spark.implicits._
    val thr = args(0).toInt
    val inputs = args(1)
    val output = args(2)


    val start_time = System.currentTimeMillis()

    // Load and preprocess the data
    val csv_rdd = sc.textFile(inputs)
    val header = csv_rdd.first()
    val data_nh = csv_rdd.filter(_ != header).map(_.split(","))
      .map(attributes => (attributes(0), Set(attributes(1))))
      .reduceByKey(_ union _)

    val data_pre = data_nh.filter(_._2.size >= thr).collectAsMap()
    val uni_user = data_pre.keys.toList
    val combuser = uni_user.combinations(2).toList

    val filter_p = sc.parallelize(combuser).filter(users => data_pre(users(0)).intersect(data_pre(users(1))).size >= thr)
    val edges_n = filter_p.flatMap(users => Seq((users(1), users(0)), (users(0), users(1))))
    val nodes = filter_p.flatMap(users => Seq(users(0), users(1))).distinct().map(user => (user))

    // Create vertices and edges DataFrames
    val vertices = nodes.toDF("id")
    val edges = edges_n.toDF("src", "dst")
    val g = GraphFrame(vertices, edges)

    // Apply the Label Propagation algorithm
    val result = g.labelPropagation.maxIter(5).run()
    val finalResult = result.rdd
      .map(row => (row.getAs[Long]("label"), Array(row.getAs[String]("id"))))
      .reduceByKey(_ ++ _)
      .map { case (label, ids) => (ids.sorted, ids.length) }
      .sortBy { case (ids, size) => (size, ids(0)) }

    val fin = finalResult.map(_._1)

    // Write the communities to a text file
    val writer = new PrintWriter(new File(output))
    fin.collect().foreach(f => {
      val com = f.map(u => s"'$u'").mkString(", ")
      writer.write(com + "\n")
    })
    writer.close()

    // Stop SparkSession
    spark.stop()
    sc.stop()

    val end_time = System.currentTimeMillis()
    val execution = (end_time - start_time) / 1000.0
    println(s"Duration: $execution")

  }
}
